c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -234,6 +238,28 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                     Some(Callee::Other { .. }) => {
                         // TODO
                     }
+                    Some(Callee::Malloc) => {
+                        // TODO
+                    }
+                    Some(Callee::Calloc) => {
+                        // TODO
+                    }
+                    Some(Callee::Realloc) => {
+                        // We handle this like a pointer assignment.
```

Comment:
What does this mean?

Reply:
the permissions of the pointer used as an argument to `realloc` are also applied to its return value. although, @spernsteiner, is this the correct behavior?

Reply:
Ah, that makes more sense.  Though I would re-phrase the comment as "We handle the pointer arg and returned pointer as a pointer assignment." or something similar.

Reply:
>is this the correct behavior?

`realloc` potentially invalidates the incoming pointer, so I think we need to treat it as a combined `free` and `malloc`.  Since the data inside the allocation is either preserved / copied over, we also need to unify non-outermost pointer types, similar to an assignment.

For a `realloc` that takes a `*mut /*1*/ *mut /*2*/ i32` and returns a `*mut /*3*/ *mut /*4*/ i32`, `PointerId(2)` and `PointerId(4)` (the non-outermost pointer types) should be unified, `PointerId(1)` should be treated like an argument to `free` (probably requiring `FREE` permission), and `PointerId(3)` should be treated like the return value of a `malloc` (which I don't think requires any special handling).

Reply:
For borrowck purposes, I think `realloc` should be treated as a combined `free` and `malloc`.  I don't think `malloc` needs any special handling (it returns a fresh `UNIQUE` pointer that doesn't alias anything).  And I'm currently thinking that `free` doesn't need any special handling either (we could have it require `UNIQUE` / treat it as a mutable access, but I think `UNIQUE` is not a strong enough property to be useful here).

I'm not totally sure about this approach, so please add a few test cases that use malloc/realloc/free (with assertions about the pointer permissions, along the lines of `tests/filecheck/alias1.rs`), and we can check if they produce reasonable results.  I'd particularly like to see how it works on this "growable vector" pattern:
```C
for (...) {
    if (len == capacity) {
        capacity *= 2;
        ptr = realloc(ptr, capacity);
    }
    // Add a new element; increase len
}
```

Reply:
IIRC, casts to and from `void*` are still not well supported, so for testing purposes, I think it would be reasonable to declare some specialized `malloc` functions in the test case:
```Rust
extern "C" {
    fn c2rust_test_typed_malloc(len: usize) -> *mut i32;
    fn c2rust_test_typed_realloc(ptr: *mut i32, len: usize) -> *mut i32;
    fn c2rust_test_typed_free(ptr: *mut i32);
}
```
And modify `util.rs` to recognize these as `Callee::Malloc` and such.

Reply:
Note that `realloc(NULL, size)` is also pretty much equivalent to `malloc(size)`, so we need to handle that case, too.

Reply:
@spernsteiner would you mind letting me know if these new tests in d5cda038 capture what you expect?

Reply:
Yes, those tests look right to me.  I'm not sure why the `realloc` test has `i` as non-`UNIQUE`, but we can investigate and fix that separately.